bode/bode/models/tag.py

```diff
@@ -0,0 +1,45 @@
+import uuid
+
+from sqlalchemy import ForeignKey
+from sqlalchemy.dialects.postgresql import UUID
+from sqlalchemy.exc import NoResultFound
+
+from bode.app import db
+
+
+class Tag(db.Model):
+    __tablename__ = "tags"
+
+    id = db.Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
+    task_id = db.Column(
+        UUID(as_uuid=True), ForeignKey("tasks.id"), nullable=False
+    )
+    name = db.Column(db.String(80), unique=True, nullable=False)
+
+    task = db.relationship("Task", back_populates="tags")
+
+    __table_args__ = (
+        db.UniqueConstraint("task_id", "name", name="_task_id_name_uc"),
+    )
```

Comment:
Something is off with this model definition. Each task can have many tags, but tag names are unique so ...you can't have two tasks with the same tag name. Kinda defeats the purpose of tags 😆. Tags are a typical many-to-many relationship so I don't see a point in trying to force them into N-1 relationship. A model like the one below should be perfectly fine for our needs at the moment:
```
tasks:
has many tasks

tags:
id - PK UUID
name - unique string
belongs to many tasks

tasks_tags:
id - PK UUID
tag_id - foreign key with index
task_id - foreign key with index
belongs to tag
belongs to task
```
[An association table without a model](https://docs.sqlalchemy.org/en/14/orm/basic_relationships.html#many-to-many) should be good enough but you can also go the [association object path](https://docs.sqlalchemy.org/en/14/orm/basic_relationships.html#association-object)

Reply:
Tag names are not unique. The pair `id`-`name` is unique so you cannot assign the same tag to the same task.
The tags can be identified by a name and they won't need any more fields so I think an additional table just for an association `name`-`tag_id` is unnecessary atm.

Reply:
What do you mean they're not unique?
```python
# models/tag.py line 17
name = db.Column(db.String(80), unique=True, nullable=False)

# migration file line 32
sa.UniqueConstraint("name"),
```

They should be unique, but with that schema, they can't be. It doesn't matter if they don't need any more fields atm, breaking out of normalized schema without any major benefits is just not the way to go. Dr Zygmunt would be angry 😆 . It makes it harder to query available tags from the DB because you need to account for the uniqueness of names when querying. It also doesn't allow to have a tag without any task assigned to it - which is a totally possible and expected use case. With DB schemas it's just easier to start with a bit more powerful one than adjust it and much of the code related to it later.

Reply:
You're right. I'll change that when I have time.
> Dr Zygmunt would be angry :laughing:
I wouldn't take that as an argument if it worked :laughing: